src/Magnum/DartIntegration/Object.h

```diff
@@ -0,0 +1,183 @@
+`BodyNode` or `ShapeNode` by passing a pointer to its constructor:
+
+@code{.cpp}
+dart::dynamics::BodyNode* body = getBodyNodeFromDart();
+SceneGraph::Object<SceneGraph::MatrixTransformation3D> object;
+DartIntegration::Object* obj = new Object{&object, body};
+@endcode
+
+or
+
+@code{.cpp}
+dart::dynamics::ShapeNode* node = getShapeNodeFromDart();
+SceneGraph::Object<SceneGraph::MatrixTransformation3D> object;
+DartIntegration::Object* obj = new Object{&object, node};
+@endcode
+
+Only the DART body/node can affect the transformation of the Magnum object and
+not the other way around. To get the latest DART transformation, you should
+update the object with @ref update().
+
+@brief convertShapeNode
+
+Convert `ShapeNode` to mesh and material data
+
+Returns false if the shape of given `ShapeNode` is
+not supported. The following DART shapes are supported:
+
+-   `BoxShape`
+-   `CapsuleShape`
+-   `CylinderShape`
+-   `EllipsoidShape`
+-   `MeshShape`
+-   `SoftMeshShape`
+-   `SphereShape`
+
+The following DART shapes are not yet supported:
+
+-   `ConeShape`
+-   `LineSegmentShape`
+-   `MultiSphereConvexHullShape`
+-   `PlaneShape` (this is an infinite plane with normal)
+
+@experimental
+*/
+class MAGNUM_DARTINTEGRATION_EXPORT Object: public SceneGraph::AbstractBasicFeature3D<Float> {
+    public:
+        /**
+         * @brief Constructor
+         * @param object    SceneGraph::Object this @ref DartIntegration::Object belongs to
+         * @param node      DART `ShapeNode` to connect with
+         */
+        template<class T> Object(T& object, dart::dynamics::ShapeNode* node = nullptr): Object{object, object, node, nullptr} {}
+
+        /**
+         * @brief Constructor
+         * @param object    SceneGraph::Object this @ref DartIntegration::Object belongs to
+         * @param body      DART `BodyNode` to connect with
+         */
+        template<class T> Object(T& object, dart::dynamics::BodyNode* body = nullptr): Object{object, object, nullptr, body} {}
+
+        /** @brief Get transformation from DART */
+        Object& update();
+
+        /** @brief Get whether Object was updated */
+        bool used();
+
+        /** @brief Clear usage flag (i.e., set it to false) */
+        Object& clearUsed();
+
+        /** @brief Get whether Object's mesh was updated */
+        bool updatedMesh();
+
+        /** @brief Get ShapeData */
+        std::reference_wrapper<ShapeData> shapeData();
+
+        /** @brief Underlying DART `ShapeNode` */
+        dart::dynamics::ShapeNode* shapeNode() { return _node; }
+
+        /** @brief Underlying DART `BodyNode` */
+        dart::dynamics::BodyNode* bodyNode() { return _body; }
+
+    private:
+        explicit Object(SceneGraph::AbstractBasicObject3D<Float>& object, SceneGraph::AbstractBasicTranslationRotation3D<Float>& transformation, dart::dynamics::ShapeNode* node, dart::dynamics::BodyNode* body);
+
+        bool convertShapeNode();
```

Comment:
So, this comment is why the review took so long (sorry about that): I tried to come up with a solution to make this whole thing more testable because otherwise it's not really possible to verify that everything is working correctly -- all you get now are some opaque Buffer and Mesh structures. I think I have it now. So, what about this: 

Previously, there was a `convertShapeNode()` function that didn't depend on GL context being present and which just returned the populated `ShapeData`. This function was very testable, so I'd vote for restoring it (together with the header it was), but with an updated API:

- it takes the `ShapeNode&`, as previously, but has no overload for `DartIntegration::Object`, as this class handles it internally
- it takes an [EnumSet](http://doc.magnum.graphics/corrade/classCorrade_1_1Containers_1_1EnumSet.html) to describe which data should be extracted and which not so the Object can update only what's needed
- it takes an existing instance of plugin manager instead of creating an instance of it inside

The Object would call `convertShapeNode()` inside, asking for extraction of only the data that changed, and then it calls `MeshTools::compile()` to populate its own internal mesh, buffer etc. instances. (So, yes, the `ShapeData` would contain just the data again.)

Instead of having the plugin manager as `static` variable inside a function, the instance would be part of the World instance and every Object would have access to it. (Reason: `static` variables inside functions are *very evil* -- they cause the compiler to generate thread-safe wrapper around these even if you aren't involve with any threading, the destruction of the whole variable is at some unspecified point after exiting from `main()`, etc. etc.)

This way the majority of the code could be properly tested again (maxing out the code coverage, verifying how the imported mesh data actually look etc.). Besides that, this would make this library less tightly integrated with SceneGraph, allowing users to implement the transformation handling in their own way, if they want to.

I hope I didn't miss anything that would make this impossible. As always, I'm open for further discussion :)

Reply:
> Previously, there was a convertShapeNode() function that didn't depend on GL context being present and which just returned the populated ShapeData. This function was very testable, so I'd vote for restoring it (together with the header it was), but with an updated API:
> 
> it takes the ShapeNode&, as previously, but has no overload for DartIntegration::Object, as this class handles it internally
> it takes an EnumSet to describe which data should be extracted and which not so the Object can update only what's needed
> it takes an existing instance of plugin manager instead of creating an instance of it inside
> The Object would call convertShapeNode() inside, asking for extraction of only the data that changed, and then it calls MeshTools::compile() to populate its own internal mesh, buffer etc. instances. (So, yes, the ShapeData would contain just the data again.)

Thanks for the comments! I am working on this and I should have something by tomorrow. I do not see any reason for now why this would not work...